434 autoregressive model and more generic feature applicator

BenoitHardier opened the pull request:
close #434

JanMaartenvanDoorn commented on openstef/data_classes/data_prep.py:
```diff
@@ -0,0 +1,99 @@
+        else:
+            raise AttributeError(f"{key} not an attribute of prediction job.")
+
+    def _load_klass(
```
```suggestion
    def _load_class(
```

JanMaartenvanDoorn commented on openstef/model/regressors/arima.py:
```diff
@@ -0,0 +1,163 @@
+class ARIMAOpenstfRegressor(OpenstfRegressor):
+    """Wrapper around statmodels implementation of (S)ARIMA(X) model.
+
+    To make prediction, it needs to update its historic, applying the fitted parameters to new data unrelated to the
```
```suggestion
    To make a prediction, it needs to update its historic, applying the fitted parameters to new data unrelated to the
```

JanMaartenvanDoorn replied:
I do not understand the word historic here, do you mean historic data?

BenoitHardier replied:
Yes I enhanced the documentation, it should be clearer.

JanMaartenvanDoorn commented on openstef/model/regressors/arima.py:
```diff
@@ -0,0 +1,163 @@
+        self.trend = trend
+
+    def fit(self, x, y, **kwargs):
+        exog = x
```
Maybe add a link to https://www.statsmodels.org/stable/endog_exog.html to give a bit more background on the exog terminology as it is not that common.

BenoitHardier replied:
Great suggestion: 
- I added multiple links to relevant documentations of statsmodels
- I removed the renaming of variables and explicitly passed the arguments as keyword argument to be clearer.

JanMaartenvanDoorn commented on openstef/model/regressors/arima.py:
```diff
@@ -0,0 +1,163 @@
+        self.feature_in_names_ = list(x.columns)
+        return self
+
+    def update_historic(self, y_past, x_past):
```
Also here I do not understand the historic terminology, do you mean history or maybe historic data?

JanMaartenvanDoorn commented on openstef/model/regressors/arima.py:
```diff
@@ -0,0 +1,163 @@
+            freq
+        )
+        n_splits = (x.shape[0] // test_size) - 1
+        tscv = TimeSeriesSplit(n_splits=n_splits, test_size=test_size)
```
Please use the full term here and not the abbreviation. I think you mean time_series_cross_validation ?

JanMaartenvanDoorn commented on openstef/model/regressors/arima.py:
```diff
@@ -0,0 +1,163 @@
+            arima_results = self.results_.apply(
+                y.iloc[apply_index], x.iloc[apply_index]
+            )
+            x_ts = x.iloc[test_index]
```
Same comment about the _ts postfix, please use the full term

BenoitHardier replied:
I refactored this part to be more readable. Names are more explicit.

JanMaartenvanDoorn reviewed (changes requested):
Looks good, nice to have ARIMA available in openSTEF! I have a few comments regarding readability.

BenoitHardier commented:
> Looks good, nice to have ARIMA available in openSTEF! I have a few comments regarding readability.

Thank you for the remarks, I refactored a bit the class and added more documentation to enhance the readability.

JanMaartenvanDoorn reviewed (approved):
Thanks for adding the documentation and updating the names, this makes it more clear!
